Acrobot swing up using trajectory optimisation

pangtao22 opened the pull request:
1. Parameters of AcrobotPlant can now be assigned in constructor. 
2. Adds an executable that creates the swing up trajectory and displays it in DrakeVisualizer. 

@hongkai-dai for feature review.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5093)
<!-- Reviewable:end -->

jwnimmer-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcPpu0tRQ1MvpegVoDw:-KcPpu0tRQ1MvpegVoDx:b-j4o17e) (no related file):*
A pile of doubles is a fairly user-unfriendly way to expose parameters, both for readability, and for the analysis toolbox.

For a preferred mechanism, see for example https://github.com/RobotLocomotion/drake/blob/master/drake/automotive/simple_car_config.named_vector and https://github.com/RobotLocomotion/drake/blob/89a02be89c1e67b30dbb0ee612d529fb408939dd/drake/automotive/simple_car.h#L62-L63 and https://github.com/RobotLocomotion/drake/blob/89a02be89c1e67b30dbb0ee612d529fb408939dd/drake/automotive/simple_car.cc#L154-L176 and https://github.com/RobotLocomotion/drake/blob/master/drake/automotive/simple_car_gen.sh.

Two notes about these citations:
(1) The `SimpleCar::SetDefaultParameters` overload between instance and static method is not required in general; the instance method alone is good enough for most cases.
(2) The `*.named_vector` parameter declaration file should have real descriptions, not just repeat the names like simple_car_config does; we haven't quite refactored the simple_car code to do that yet (it's under active work).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cpp, line 1 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcPsvCrqc8xVQkyUgEk:-KcPsvCsEKCcAW-mv6r6:b-gl0wuz) ([raw file](https://github.com/robotlocomotion/drake/blob/f78b8da2871256b9020bf66076ec32abde7cbb73/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cpp#L1)):*
> ```cpp
> // Generates a swing-up trajectory for acrobot and displays the trajectory
> ```

use extension `.cc` instead of `.cpp`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcPpu0tRQ1MvpegVoDw:-KcQ8_8_lNswYyShjx-R:b-5vfoac) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

A pile of doubles is a fairly user-unfriendly way to expose parameters, both for readability, and for the analysis toolbox.

For a preferred mechanism, see for example https://github.com/RobotLocomotion/drake/blob/master/drake/automotive/simple_car_config.named_vector and https://github.com/RobotLocomotion/drake/blob/89a02be89c1e67b30dbb0ee612d529fb408939dd/drake/automotive/simple_car.h#L62-L63 and https://github.com/RobotLocomotion/drake/blob/89a02be89c1e67b30dbb0ee612d529fb408939dd/drake/automotive/simple_car.cc#L154-L176 and https://github.com/RobotLocomotion/drake/blob/master/drake/automotive/simple_car_gen.sh.

Two notes about these citations:
(1) The `SimpleCar::SetDefaultParameters` overload between instance and static method is not required in general; the instance method alone is good enough for most cases.
(2) The `*.named_vector` parameter declaration file should have real descriptions, not just repeat the names like simple_car_config does; we haven't quite refactored the simple_car code to do that yet (it's under active work).
</blockquote></details>

To be clear, having not actually done a code review pass here I'm not intending to post my above as a defect, merely a suggestion -- and mostly just to get the citations out there for the actual author and reviewers to consider.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcPpu0tRQ1MvpegVoDw:-KcQIheHAYCIGdRPWzO1:bfkk9y8) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

To be clear, having not actually done a code review pass here I'm not intending to post my above as a defect, merely a suggestion -- and mostly just to get the citations out there for the actual author and reviewers to consider.
</blockquote></details>

Thanks for the suggestion! There has been a //TODO for Russ to include the parameters in the Context for some time. Now is probably a good time to find a better place to put the parameters. I'll look into it. 

---

*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cpp, line 1 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcPsvCrqc8xVQkyUgEk:-KcQIfhWzD75EJfzLxwn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f78b8da2871256b9020bf66076ec32abde7cbb73/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cpp#L1)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

use extension `.cc` instead of `.cpp`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 2 of 8 files at r1.
Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcTyzBRpXmiFXRIQagN:-KcTyzBRpXmiFXRIQagO:b-mnekka) ([raw file](https://github.com/robotlocomotion/drake/blob/ffc7bf2d322a57e1774eb808c091b0e9b0dd5e39/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc#L47)):*
> ```cpp
> 
>   const int kNumTimeSamples = 21;
>   const int kTrajectoryTimeLowerBound = 2;
> ```

const double instead of const int, same for the upper bound.

---

*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcTz9F0TDlRjUixkJPt:-KcTz9F0TDlRjUixkJPu:bn7ly7t) ([raw file](https://github.com/robotlocomotion/drake/blob/ffc7bf2d322a57e1774eb808c091b0e9b0dd5e39/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc#L91)):*
> ```cpp
>   simulator.set_target_realtime_rate(FLAGS_realtime_factor);
>   simulator.Initialize();
>   simulator.StepTo(kTrajectoryTimeUpperBound);
> ```

If the simulator steps to the time upper bound, and `pp_xtraj` has a shorter time span, what will be the input after the `pp_xtraj` ends?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
+@hongkai-dai for feature review. @pangtao22 , when adding people as the reviewer, you need to add `+` in front of the `@` sign.

---

Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-KcYv-_2VBq9lr3I5KJw:b-85lgv6)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Sorry! But I do not have the permission to add assignees. This is the message I got when I try to use + sign: 

Failed to publish: Sorry, you don't have permission to set labels or assignees in this repository.

---

Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcTz9F0TDlRjUixkJPt:-KcZRF6iifWMj_UhQyNI:b-s6ysv1) ([raw file](https://github.com/robotlocomotion/drake/blob/ffc7bf2d322a57e1774eb808c091b0e9b0dd5e39/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc#L91)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

If the simulator steps to the time upper bound, and `pp_xtraj` has a shorter time span, what will be the input after the `pp_xtraj` ends?
</blockquote></details>

I actually do not know... In the simulation, the robot stays at the upright position. Is there a way to log the input of the trajectory source?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-KcZQLRsfOOY30wfDHJI:bagqy1g)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcTz9F0TDlRjUixkJPt:-KcZ_oqYSKQlmp2VTPVv:b-7v8n5v) ([raw file](https://github.com/robotlocomotion/drake/blob/ffc7bf2d322a57e1774eb808c091b0e9b0dd5e39/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc#L91)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

I actually do not know... In the simulation, the robot stays at the upright position. Is there a way to log the input of the trajectory source?
</blockquote></details>

@liangfok @siyuanfeng-tri  do you know how to log the input ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcTz9F0TDlRjUixkJPt:-KcZaPtEJB47s_DOnZje:b-tjgzxb) ([raw file](https://github.com/robotlocomotion/drake/blob/ffc7bf2d322a57e1774eb808c091b0e9b0dd5e39/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc#L91)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

@liangfok @siyuanfeng-tri  do you know how to log the input ?
</blockquote></details>

@hongkai-dai @pangtao22 you can use the SignalLogger in systems/primitives to do this, after simulation, you can get a flat array of data vs time. 

for interp pp, t is clamped between the start and the end. so if you run past the end, the output stays the same as the last tick. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
@russtedrake , could you add @pangtao22 to the list of developers, so that he can set labels or assignees?

---

Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-KcZzRxQlNlDmKvbg3uD:bwxl2oe)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot linux-xenial-gcc-experimental-open-source please
@drake-jenkins-bot linux-xenial-cpplint-experimental please

david-german-tri commented:
> @russtedrake , could you add @pangtao22 to the list of developers, so that he can set labels or assignees?

Done.

pangtao22 commented:
Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcTyzBRpXmiFXRIQagN:-Kce2hve32_xb8xYwjjg:blp0x3a) ([raw file](https://github.com/robotlocomotion/drake/blob/ffc7bf2d322a57e1774eb808c091b0e9b0dd5e39/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc#L47)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

const double instead of const int, same for the upper bound.
</blockquote></details>

Will push this change together with other changes. 

---

*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcTz9F0TDlRjUixkJPt:-Kce2UJhqyrwKnT01M1p:b-gk1bh6) ([raw file](https://github.com/robotlocomotion/drake/blob/ffc7bf2d322a57e1774eb808c091b0e9b0dd5e39/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc#L91)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

@hongkai-dai @pangtao22 you can use the SignalLogger in systems/primitives to do this, after simulation, you can get a flat array of data vs time. 

for interp pp, t is clamped between the start and the end. so if you run past the end, the output stays the same as the last tick. 
</blockquote></details>

@hongkai-dai, do you think I need change the current code?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcTz9F0TDlRjUixkJPt:-Kce5nXBq-tiZf4x3gbA:bprpi53) ([raw file](https://github.com/robotlocomotion/drake/blob/ffc7bf2d322a57e1774eb808c091b0e9b0dd5e39/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc#L91)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

@hongkai-dai, do you think I need change the current code?
</blockquote></details>

I am satisfied with the current code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KcTyzBRpXmiFXRIQagN:-KceCkVZ-zYtX-_dHagt:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ffc7bf2d322a57e1774eb808c091b0e9b0dd5e39/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc#L47)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

Will push this change together with other changes. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_swing_up.cc, line 80 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KceaTbCEIBE2hEISytr:-KceaTbD2ZZbdpKQR7Ji:b-tojknh) ([raw file](https://github.com/robotlocomotion/drake/blob/f72ac68e4b9f1f3509b0419cd375a9b9db534bdb/drake/examples/Acrobot/acrobot_swing_up.cc#L80)):*
> ```cpp
> 
>   // Current limit for MIT's acrobot is 7-9 Amps, accroding to Michael Posa.
>   const int kTorqueLimit = 8;
> ```

const double kTorqueLimit

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_swing_up.cc, line 80 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KceaTbCEIBE2hEISytr:-KcejjeLhlUmkODfFVr-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f72ac68e4b9f1f3509b0419cd375a9b9db534bdb/drake/examples/Acrobot/acrobot_swing_up.cc#L80)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

const double kTorqueLimit
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> +@sherm1 +@soonho-tri for platform review.

---

Review status: 2 of 6 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-KcelrW569jcyJCpGmA5:b-k7odu3)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 2 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-Kchq4v7siVdJfBsyEdv:-Kchq4v7siVdJfBsyEdw:b-c9llqg) (no related file):*
This one doesn't change bazel `BUILD` file. Is it the case that this example/test only included in CMake build?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 3 of 8 files at r1, 3 of 5 files at r2, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_plant.h, line 73 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KchriigykW0DpVl-onB:-KchriigykW0DpVl-onC:ba1bwmi) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_plant.h#L73)):*
From Google's C++ Guide: https://google.github.io/styleguide/cppguide.html#Copyable_Movable_Types

> If you do not want to support copy/move operations on your type, explicitly disable them

I guess this class is non-movable? If so, please  consider using `DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN`.

---

*[drake/examples/Acrobot/acrobot_swing_up.cc, line 12 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-Kchtw_KsEqYDOAMfc7Z:-Kchtw_L_nc8CS--cjyx:b-3zbh8x) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_swing_up.cc#L12)):*
> ```cpp
> #include "drake/common/eigen_types.h"
> #include "drake/examples/Acrobot/gen/acrobot_state_vector.h"
> #include "drake/examples/Acrobot/acrobot_swing_up.h"
> ```

This `#include` should be atop the file.

---

*[drake/examples/Acrobot/acrobot_swing_up.cc, line 15 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-Kchu9f-1-8d53MQLL3e:-Kchu9f-1-8d53MQLL3f:b-dut7bq) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_swing_up.cc#L15)):*
> ```cpp
> #include "drake/examples/Acrobot/acrobot_plant.h"
> #include "drake/solvers/function.h"
> #include "drake/systems/trajectory_optimization/direct_collocation.h"
> ```

Please consider sorting the `#include`s block.

---

*[drake/examples/Acrobot/CMakeLists.txt, line 18 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KchvuU0CdW6_kXxIdT_:-KchvuU1pF7fX6LpfQ1k:bb0p0mt) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/CMakeLists.txt#L18)):*
> ```CMake
>         drake-trajectory-optimization
>         eigen3)
> ```

The above changes look good to me. But I think it's better to have cmake expert's confirmation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform review -- a few comments, PTAL. Also I would like to know that a feature reviewer has run the example and it looks reasonable (@hongkai-dai volunteered).

---

Reviewed 2 of 8 files at r1, 3 of 5 files at r2, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_plant.h, line 27 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KctfNk0VFm6_OBmu8qS:-KctfNk2bAIGp99n6F3l:b-953qbi) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_plant.h#L27)):*
> ```objc
> /// @param lc1 Vertical distance from shoulder joint to center of mass of
> /// link 1 (m).
> /// @param lc2 Vertical distance from elbox joint to center of mass of
> ```

BTW, elbox->elbow

---

*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc, line 44 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-Kctg-XpLmmtL8HukI-2:-Kctg-XqcnWbZ6UHvZBc:b-tfnwl5) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc#L44)):*
> ```cpp
>   auto acrobot_ptr = AcrobotPlant<double>::CreateAcrobotMIT();
>   acrobot = acrobot_ptr.get();
>   DRAKE_DEMAND(acrobot != nullptr);
> ```

You don't need a raw pointer here. Instead, just work with the returned unique ptr:
```c++
auto acrobot = AcrobotPlant<double>::CreateAcrobotMIT();
```
You can then use `acrobot->` or `*acrobot` for dereferencing exactly as you did with the raw pointer, and `acrobot != nullptr` works as expected. The only change below is to use `acrobot.get()` when you need mutable access to the raw pointer.

---

*[drake/examples/Acrobot/acrobot_swing_up.cc, line 74 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-Kctp7K1lG5SkJT1o7KY:-Kctp7K1lG5SkJT1o7KZ:bv4kaya) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_swing_up.cc#L74)):*
> ```cpp
> }  // anon namespace
> 
> void AddSwingUpTrajectoryParams(
> ```

BTW, does this have to be a free function? Would it make sense to make it a member of AcrobotPlant instead? (I'm not sure that's better, just wanted to mention it in case it makes sesne. It is a little awkward to use a free function here IMO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-Kctpq_usDRj6IE_p6hR:b8ysk7)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_plant.cc, line 51 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KczSuEcflSIBV4BM9di:-KczSuEcflSIBV4BM9dj:b2mfbsp) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_plant.cc#L51)):*
> ```cpp
>                       0.2563,   // l1
>                       0,        // l2
>                       1.6738,   // lc1
> ```

BTW, odd indentation. Is this what `clang-format` did?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: 5 of 9 files reviewed at latest revision, 10 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-Kchq4v7siVdJfBsyEdv:-KdEXfl86Gx8vQjIirsj:b-nqfn1v) (no related file):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

This one doesn't change bazel `BUILD` file. Is it the case that this example/test only included in CMake build?
</blockquote></details>

Added bazel build for the executables listed below and their relevant libraries
acrobot_plant_with_lcm
acrobot_spong_controller_w_lcm
acrobot_run_swing_up_traj_optimization
acrobot_run_swing_up

---

*[drake/examples/Acrobot/acrobot_plant.cc, line 51 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KczSuEcflSIBV4BM9di:-KdEYOVkggK2lVAV5HCz:bmyh67l) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_plant.cc#L51)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, odd indentation. Is this what `clang-format` did?
</blockquote></details>

Ran `clang-format` again on this file and indentation looks normal now. 

---

*[drake/examples/Acrobot/acrobot_plant.h, line 27 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KctfNk0VFm6_OBmu8qS:-KdEYgFSKXTCJLjwv4MB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_plant.h#L27)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, elbox->elbow
</blockquote></details>

Done.

---

*[drake/examples/Acrobot/acrobot_plant.h, line 73 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KchriigykW0DpVl-onB:-KdNAEVjyTwX-M1PwbvJ:bs7wumb) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_plant.h#L73)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

From Google's C++ Guide: https://google.github.io/styleguide/cppguide.html#Copyable_Movable_Types

> If you do not want to support copy/move operations on your type, explicitly disable them

I guess this class is non-movable? If so, please  consider using `DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN`.

</blockquote></details>

Done.

---

*[drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc, line 44 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-Kctg-XpLmmtL8HukI-2:-KdE_AJqdeps27hTCVtt:bdzvuos) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_run_swing_up_traj_optimization.cc#L44)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

You don't need a raw pointer here. Instead, just work with the returned unique ptr:
```c++
auto acrobot = AcrobotPlant<double>::CreateAcrobotMIT();
```
You can then use `acrobot->` or `*acrobot` for dereferencing exactly as you did with the raw pointer, and `acrobot != nullptr` works as expected. The only change below is to use `acrobot.get()` when you need mutable access to the raw pointer.
</blockquote></details>

Changed the code as advised​. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: 5 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/examples/Acrobot/acrobot_swing_up.cc, line 12 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-Kchtw_KsEqYDOAMfc7Z:-KdNArnLmnVxvv-Ss2vI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_swing_up.cc#L12)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

This `#include` should be atop the file.
</blockquote></details>

Done.

---

*[drake/examples/Acrobot/acrobot_swing_up.cc, line 15 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-Kchu9f-1-8d53MQLL3e:-KdNAsSPLDzBroxv0UmH:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_swing_up.cc#L15)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Please consider sorting the `#include`s block.
</blockquote></details>

Done.

---

*[drake/examples/Acrobot/acrobot_swing_up.cc, line 74 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-Kctp7K1lG5SkJT1o7KY:-KdNBIAudzXLTWebA991:b-a2od62) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/acrobot_swing_up.cc#L74)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, does this have to be a free function? Would it make sense to make it a member of AcrobotPlant instead? (I'm not sure that's better, just wanted to mention it in case it makes sesne. It is a little awkward to use a free function here IMO.
</blockquote></details>

This is how this function is implemented in `pendulum_swing_up.h`. I am not sure about the best way to put this function either.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 8 files at r1, 3 of 5 files at r2, 5 of 7 files at r5.
Review status: 7 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/Acrobot/BUILD, line 75 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KdOq1vAlQlt0_Rmm26X:-KdOppf-uv7S9OZm8PEx:b-e6rn1) ([raw file](https://github.com/robotlocomotion/drake/blob/1702e791ce8ef35c0b4ef4fc9e715a16b4201ce8/drake/examples/Acrobot/BUILD#L75)):*
> ```Python
>         "//drake/lcm:lcm",
>         "//drake/multibody/rigid_body_plant:drake_visualizer",
>         "@gflags//:gflags",
> ```

Why you need gflags?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: 7 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/Acrobot/BUILD, line 75 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KdOq1vAlQlt0_Rmm26X:-KdRNR4e4mPfoW0t1CgS:bm2dkg8) ([raw file](https://github.com/robotlocomotion/drake/blob/1702e791ce8ef35c0b4ef4fc9e715a16b4201ce8/drake/examples/Acrobot/BUILD#L75)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Why you need gflags?
</blockquote></details>

To set the real-time factor (used at https://github.com/RobotLocomotion/drake/pull/5093/files#diff-4304cda77fb15ea52e0ce2e660a669beR84)? I am using it because some earlier acrobot examples (such as https://github.com/RobotLocomotion/drake/blob/master/drake/examples/Acrobot/acrobot_run_lqr.cc) are also using gflags to set the real-time factor. 

---

*[drake/examples/Acrobot/CMakeLists.txt, line 18 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KchvuU0CdW6_kXxIdT_:-KdRO7WdC0fLtxBqYMn8:b-x60c8s) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/CMakeLists.txt#L18)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

The above changes look good to me. But I think it's better to have cmake expert's confirmation.
</blockquote></details>

Could you ask some cmake expert to take a look? I am not familiar with different people's expertise. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 7 of 9 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/Acrobot/CMakeLists.txt, line 18 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KchvuU0CdW6_kXxIdT_:-KdT3aM-zHFxj3cuReVn:bv0wzbp) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/CMakeLists.txt#L18)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

Could you ask some cmake expert to take a look? I am not familiar with different people's expertise. Thanks!
</blockquote></details>

Do we foresee downstream users of Drake wanting to link against `libdrakeAcrobotPlant`? To my knowledge, that's the main reason for installing the header and pkg_config file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 7 of 9 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/Acrobot/CMakeLists.txt, line 18 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KchvuU0CdW6_kXxIdT_:-KdT4Dox6mmgtqkjiCYE:b2qtb44) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/CMakeLists.txt#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Do we foresee downstream users of Drake wanting to link against `libdrakeAcrobotPlant`? To my knowledge, that's the main reason for installing the header and pkg_config file.
</blockquote></details>

it's possible, yes.  and we export the analogous pendulum plant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 7 of 9 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/Acrobot/CMakeLists.txt, line 18 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KchvuU0CdW6_kXxIdT_:-KdT5Fx6CwhI7Ca-u45P:bl832vy) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/CMakeLists.txt#L18)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

it's possible, yes.  and we export the analogous pendulum plant.
</blockquote></details>

Alright, then this change looks good to me, but I'll defer to a real `CMake` expert like @mwoehlke-kitware or @jamiesnape.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 7 of 9 files reviewed at latest revision, 4 unresolved discussions.

---

*[tools/gfortran.bzl, line 6 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KdTiGlLeP3XUjasPVQV:-KdTiGlLeP3XUjasPVQW:b-f2gcg) ([raw file](https://github.com/robotlocomotion/drake/blob/1702e791ce8ef35c0b4ef4fc9e715a16b4201ce8/tools/gfortran.bzl#L6)):*
> ```Python
> 
> def _find_and_symlink(repository_ctx, lib_name):
>     gfortran = repository_ctx.which("gfortran")
> ```

FYI, this line is already added by https://github.com/RobotLocomotion/drake/commit/dbf3cdb792cc671d2854a9cbad97aa648fe5872f

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

---

Reviewed 2 of 5 files at r2, 7 of 7 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/Acrobot/BUILD, line 59 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KdYG7mb0CRV-1f0M3xi:-KdYG7mb0CRV-1f0M3xj:bpd3fm9) ([raw file](https://github.com/robotlocomotion/drake/blob/1702e791ce8ef35c0b4ef4fc9e715a16b4201ce8/drake/examples/Acrobot/BUILD#L59)):*
> ```Python
>         "//drake/multibody/parsers",
>         "//drake/systems/sensors:rotary_encoders",
>         "//drake/systems/framework",      
> ```

BTW, trailing blanks on this line?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-KdYGTPeHezQ-qoHMvrd:b-cxwtpi)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
BTW, @hongkai-dai did run this and gives it a :+1:.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-KdYGc-yM33COe4T1DHc:b-j2a1xj)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/Acrobot/BUILD, line 59 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KdYG7mb0CRV-1f0M3xi:-KdYHrw2agVv0xJoNXzy:bcq93qn) ([raw file](https://github.com/robotlocomotion/drake/blob/1702e791ce8ef35c0b4ef4fc9e715a16b4201ce8/drake/examples/Acrobot/BUILD#L59)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, trailing blanks on this line?
</blockquote></details>

FYI, running `tools/buildifier.sh` should fix these problems.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-KdYI-THEi1DNSvKwUgm:bnfp4nl)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: 7 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/Acrobot/BUILD, line 59 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KdYG7mb0CRV-1f0M3xi:-KdbFSn-zoMCH7qKnGOR:bs7wumx) ([raw file](https://github.com/robotlocomotion/drake/blob/1702e791ce8ef35c0b4ef4fc9e715a16b4201ce8/drake/examples/Acrobot/BUILD#L59)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, running `tools/buildifier.sh` should fix these problems.
</blockquote></details>

Done. 

---

*[tools/gfortran.bzl, line 6 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KdTiGlLeP3XUjasPVQV:-KdbEK9bvhb8cV_1Xd8Q:bse51nb) ([raw file](https://github.com/robotlocomotion/drake/blob/1702e791ce8ef35c0b4ef4fc9e715a16b4201ce8/tools/gfortran.bzl#L6)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, this line is already added by https://github.com/RobotLocomotion/drake/commit/dbf3cdb792cc671d2854a9cbad97aa648fe5872f
</blockquote></details>

Merged with the latest master branch. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 2 of 5 files at r2, 5 of 7 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Acrobot/CMakeLists.txt, line 18 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KchvuU0CdW6_kXxIdT_:-KdbIpX4qnpPB6Ue-vhy:br6om41) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/CMakeLists.txt#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Alright, then this change looks good to me, but I'll defer to a real `CMake` expert like @mwoehlke-kitware or @jamiesnape.
</blockquote></details>

Should I use the + sign to assign @mwoehlke-kitware or @jamiesnape to review this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@mwoehlke-kitware for feature review of the changes to `CMakeLists.txt` please. Thanks!

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Acrobot/CMakeLists.txt, line 18 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KchvuU0CdW6_kXxIdT_:-KdbMgI5264jt9FZsH_e:bvt2q5t) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/CMakeLists.txt#L18)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

Should I use the + sign to assign @mwoehlke-kitware or @jamiesnape to review this?
</blockquote></details>

Yes, I will go ahead and do that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-KdbMilnXLw9MCqGBqy_:b-hoyvud)*
<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
@jamiesnape knows those bits better

-@mwoehlke-kitware +@jamiesnape

jamiesnape commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/Acrobot/CMakeLists.txt, line 18 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KchvuU0CdW6_kXxIdT_:-KdfOQfhiZbFT-ANWH9l:b-3c1a7i) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/CMakeLists.txt#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Yes, I will go ahead and do that.
</blockquote></details>

The changes are correct, but apparently there was already a breakage from a previous PR. Add
```
drake-linear-control
drake-multibody-parsers
drake-sensors
drake-system-framework
```
to the `REQUIRES` for `drake_install_pkg_config_file`

---

*[drake/examples/Acrobot/CMakeLists.txt, line 95 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KdfP42s-xcWHnlZ-8uZ:-KdfP42s-xcWHnlZ-8u_:bhyvods) ([raw file](https://github.com/robotlocomotion/drake/blob/be57dfee22b14f44866717cdac8a291e43bc9c71/drake/examples/Acrobot/CMakeLists.txt#L95)):*
> ```CMake
>           drakeRigidBodyPlant
>           drakeSystemAnalysis
>           GTest::GTest)
> ```

Is this a unit test? If so, use [`drake_add_cc_test(acrobot_run_swing_up_traj_optimization)`](https://github.com/RobotLocomotion/drake/blob/master/cmake/test.cmake#L93-L126) and remove `GTest::GTest` from the `target_link_libraries`.

---

*[drake/lcmtypes/BUILD, line 12 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KdfPSI5jtfKP0cwL4v4:-KdfPSI5jtfKP0cwL4v5:bp3lgla) ([raw file](https://github.com/robotlocomotion/drake/blob/be57dfee22b14f44866717cdac8a291e43bc9c71/drake/lcmtypes/BUILD#L12)):*
> ```Python
>     lcm_package = "drake",
>     lcm_srcs = ["lcmt_acrobot_u.lcm",
>                 "lcmt_acrobot_x.lcm",
> ```

BTW, not sure this indent is correct.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-KdfPefs4GY4u2Z9P7eS:bnfp4nl)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/Acrobot/CMakeLists.txt, line 18 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KchvuU0CdW6_kXxIdT_:-Ke-zQdx6DbualErkZYm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/437062c652b3b149a34b5d7b1d3748b8dbfa8dc4/drake/examples/Acrobot/CMakeLists.txt#L18)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

The changes are correct, but apparently there was already a breakage from a previous PR. Add
```
drake-linear-control
drake-multibody-parsers
drake-sensors
drake-system-framework
```
to the `REQUIRES` for `drake_install_pkg_config_file`
</blockquote></details>

Done.

---

*[drake/examples/Acrobot/CMakeLists.txt, line 95 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KdfP42s-xcWHnlZ-8uZ:-Ke-zXkcdenJA62LFm3Q:b-olc3bx) ([raw file](https://github.com/robotlocomotion/drake/blob/be57dfee22b14f44866717cdac8a291e43bc9c71/drake/examples/Acrobot/CMakeLists.txt#L95)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Is this a unit test? If so, use [`drake_add_cc_test(acrobot_run_swing_up_traj_optimization)`](https://github.com/RobotLocomotion/drake/blob/master/cmake/test.cmake#L93-L126) and remove `GTest::GTest` from the `target_link_libraries`.
</blockquote></details>

No, this is an executable, not a unit test. Removed `GTest::GTest`. 

---

*[drake/lcmtypes/BUILD, line 12 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-KdfPSI5jtfKP0cwL4v4:-Ke01I5056I2QIScVmqH:bs0uq1v) ([raw file](https://github.com/robotlocomotion/drake/blob/be57dfee22b14f44866717cdac8a291e43bc9c71/drake/lcmtypes/BUILD#L12)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

BTW, not sure this indent is correct.
</blockquote></details>

Ran buildifier. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 6 of 8 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-Ke058Kr_ju47OslLuDV:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Reviewed 1 of 8 files at r1, 2 of 5 files at r2, 4 of 7 files at r5, 1 of 2 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-Ke07iioQUjxmNLkmBIZ:b-3f2dwx)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
-(status: curate commits before merging) 

---

Review status: 6 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5093#-:-Ke0M5P4ajPZDZ0WSj_p:b-l70wzf)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
For the record, the settings added in this PR (for the MIT parameters) were not written to transmogrify (e.g., `ToAutodDiff`) correctly.
